torch/csrc/jit/ir.cpp

```diff
@@ -251,6 +256,335 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  // graph, record the value -> new generated name mapping
+  std::unordered_map<const Value*, const Value*> aliases_;
+  // std::unordered_map<const Value*, std::string> aliases_;
+
+  // The Graph already tracks unique_names_, this is just for additional ones
+  // generated during printing
+  std::unordered_map<std::string, const Value*> generated_names_;
+
+  template<class T>
+  void dualIterator(
+    at::ArrayRef<T> list_a,
+    at::ArrayRef<T> list_b,
+    const size_t offset_a,
+    const size_t offset_b,
+    std::function<void(T, T)> action
+  ) const {
+    auto it_a = list_a.begin() + offset_a;
+    auto it_b = list_b.begin() + offset_b;
+
+    for (; it_a != list_a.end() && it_b != list_b.end(); ++it_a, ++it_b) {
+      action(*it_a, *it_b);
+    }
+  }
+
+  std::ostream& printValueList(
+    std::ostream & out,
+    at::ArrayRef<const Value*> list
+  ) {
+    out << "(";
+    auto delimiter = "";
+    for (const auto* value : list) {
+      out << delimiter;
+      printValue(out, value);
+      delimiter = ", ";
+    }
+    out << ")";
+    return out;
+  }
+
+  void printAssignment(
+    std::ostream& out,
+    const Value* lhs,
+    const Value* rhs,
+    const size_t level
+  ) {
+    indent(out, level);
+    printValue(out, lhs);
+    out << " = ";
+    printValue(out, rhs);
+    out << std::endl;
+  }
+
+  std::ostream& printIf(
+    std::ostream & out,
+    const Node* node,
+    const size_t level
+  ) {
+    indent(out, level);
+    out << "if ";
+    const auto if_block = node->blocks()[0];
+    const auto else_block = node->blocks()[1];
+    printValue(out, node->inputs()[0]);
+    out << ":" << std::endl;
+
+    // Print node contents
+    printBlock(out, if_block, level + 1);
+
+    // Print if block output
+    dualIterator<const Value*>(
+      node->outputs(),
+      if_block->outputs(),
+      0, 0,
+      [&](const Value* node_output, const Value* return_input) {
+        printAssignment(out, node_output, return_input, level + 1);
+      }
+    );
+
+    indent(out, level);
+    out << "else:" << std::endl;
+    printBlock(out, else_block, level + 1);
+    dualIterator<const Value*>(
+      node->outputs(),
+      else_block->outputs(),
+      0, 0,
+      [&](const Value* node_output, const Value* return_input) {
+        printAssignment(out, node_output, return_input, level + 1);
+      }
+    );
+
+    return out;
+  }
+
+  bool isValueUsedLater(
+    const Value* val
+  ) const {
+    if (aliases_.find(val) != aliases_.end()) {
+      return true;
+    }
+    return val->uses().size() > 0;
+  }
+
+  std::ostream& printLoop(
+    std::ostream & out,
+    const Node* node,
+    const size_t level
+  ) {
+    const auto body_block = node->blocks()[0];
+    // aliases_[body_block->inputs()[0]] = body_block->inputs()[0]->uniqueName();
+    aliases_[body_block->inputs()[0]] = body_block->inputs()[0];
+
+    // The block outputs are not live after the end of the block, so we can use
```

Comment:
I don't think this is true, what happens in this case where `x` is the value of the output for loop-carried dependency `z`, but still is used after the loop?

```
def foo(y):
   x = y + 1
   z = 4
   while y < 8:
      y += 1
      z = x
  return x, z
```

I would think the safe values to use for the loop carried dependencies are the `node->outputs()` which cannot be live when entering the loop.

Reply:
> In that particular case z gets aliased to x at the end and it outputs return (x, x)

That is wrong though -- if the loop never executes the value of `z` should be `4` not `x`.

Reply:
I see. How about something like this:

```python
  def script(y1):
    x = aten::add(y1, 1, 1)
    z1 = aten::add(x, 5, 1)
    t9 = aten::lt(y1, 8)
    t10 = prim::TensorToNum(t9)
    t11 = t10
    t12 = y1
+   z = z1
    while t11:
      y2 = aten::add(t12, 1, 1)
      t17 = aten::lt(y2, 8)
      t18 = prim::TensorToNum(t17)
      t11 = t18
      t12 = y2
+     z = x
    return (x, z)
```

I think the extra assignments are necessary for correctness. A followup PR could deal with pruning the unnecessary ones / improving the naming.